[Quant][devs] Separated implementations for quantized & non-quantized tensors for the squeeze function

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71639
* #71876

Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR separates the calls to quantized & non-quantized backends for squeeze
using a dispatcher.

Test plan: Additional testing was not implemented to test this change because test cases in the existing test case already make use of the squeeze function for various backends.

initial

Differential Revision: [D33798546](https://our.internmc.facebook.com/intern/diff/D33798546)

dzdang committed:
[Quant] Separated implementations for quantized & non-quantized tensors for the squeeze function

Summary: This is a PR that address a concern brought up in https://github.com/pytorch/pytorch/issues/54150 regarding opting for separat kernels for quantized & non-quantized tensors, when possible, as opposed to using conditional if-else.

Test plan: Additional testing was not implemented to test this change because test cases in the existing test case already make use of the squeeze function for various backends.

[ghstack-poisoned]

dzdang committed:
Update on "[Quant] Separated implementations for quantized & non-quantized tensors for the squeeze function"


Summary: This is a PR that address a concern brought up in https://github.com/pytorch/pytorch/issues/54150 regarding opting for separat kernels for quantized & non-quantized tensors, when possible, as opposed to using conditional if-else.

Test plan: Additional testing was not implemented to test this change because test cases in the existing test case already make use of the squeeze function for various backends.

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1991,6 +1989,12 @@ Tensor squeeze(const Tensor& self, int64_t dim) {
   return result;
 }
 
+Tensor squeeze_quantized(const Tensor& self, int64_t dim) {
```
can we merge the two variants of squeeze_qtensor, and also merge the implementations of squeeze_quantized (call one squeeze_quantized from another, ideally we can call squeeze_quantized(const Tensor& self, int64_t dim) from squeeze_quantized(const Tensor& self))

dzdang replied:
I should be able to merge the 2 squeeze_qtensor into a single function by making the dim arg an optional arg and then make some modifications to handle each case based on whether the optional arg is present or not. Is this what you mean, or do you mean I should call squeeze_qtensor(const Tensor& self, int64_t dim) from within the for loop of squeeze_qtensor(const Tensor& self)?

Re merging squeeze_quantized:
I'll follow up with you through work chat. There might be an issue with the previous un-dispatched version

jerryzh168 replied:
> I should be able to merge the 2 squeeze_qtensor into a single function by making the dim arg an optional arg and then make some modifications to handle each case based on whether the optional arg is present or not. Is this what you mean, or do you mean I should call squeeze_qtensor(const Tensor& self, int64_t dim) from within the for loop of squeeze_qtensor(const Tensor& self)?

oh I was assuming the no-dim variant is a special case for the dim-variant. if so we can do the first things you proposed here, if not, then I feel it might be OK to keep them this way. But looking at the implementation of squeeze, it looks like there can be a lot of code sharing happening between these two variants, wondering if we can still share the code between the two squeeze_qtensor variants (currently it's duplicated).

yeah modification for squeeze_quantized will depend on how we implement squeeze_qtensor.

jerryzh168 replied:
optional argument for dim: we can do this in implementation, not in interface/signature, we still want to exactly match the fp32 signature

dzdang replied:
sorry I don't understand. are you suggesting we should make dim an optional argument in squeeze_quantized. if so, I'm confused how we can get this to work without modifying native_functions.yaml

dzdang replied:
done. can you approve?

dzdang committed:
Update on "[Quant] Separated implementations for quantized & non-quantized tensors for the squeeze function"


Summary: This is a PR that address a concern brought up in https://github.com/pytorch/pytorch/issues/54150 regarding opting for separat kernels for quantized & non-quantized tensors, when possible, as opposed to using conditional if-else.

Test plan: Additional testing was not implemented to test this change because test cases in the existing test case already make use of the squeeze function for various backends.

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33798546).

dzdang committed:
Update on "[Quant] Separated implementations for quantized & non-quantized tensors for the squeeze function"


Summary: This is a PR that address a concern brought up in https://github.com/pytorch/pytorch/issues/54150 regarding opting for separat kernels for quantized & non-quantized tensors, when possible, as opposed to using conditional if-else.

Test plan: Additional testing was not implemented to test this change because test cases in the existing test case already make use of the squeeze function for various backends.

Differential Revision: [D33798546](https://our.internmc.facebook.com/intern/diff/D33798546)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant] Separated implementations for quantized & non-quantized tensors for the squeeze function"


Summary: This is a PR that address a concern brought up in https://github.com/pytorch/pytorch/issues/54150 regarding opting for separat kernels for quantized & non-quantized tensors, when possible, as opposed to using conditional if-else.

Test plan: Additional testing was not implemented to test this change because test cases in the existing test case already make use of the squeeze function for various backends.

Differential Revision: [D33798546](https://our.internmc.facebook.com/intern/diff/D33798546)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant] Separated implementations for quantized & non-quantized tensors for the squeeze function"


Summary: This is a PR that address a concern brought up in https://github.com/pytorch/pytorch/issues/54150 regarding opting for separat kernels for quantized & non-quantized tensors, when possible, as opposed to using conditional if-else.

Test plan: Additional testing was not implemented to test this change because test cases in the existing test case already make use of the squeeze function for various backends.

Differential Revision: [D33798546](https://our.internmc.facebook.com/intern/diff/D33798546)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant] Separated implementations for quantized & non-quantized tensors for the squeeze function"


Summary: This is a PR that address a concern brought up in https://github.com/pytorch/pytorch/issues/54150 regarding opting for separat kernels for quantized & non-quantized tensors, when possible, as opposed to using conditional if-else.

Test plan: Additional testing was not implemented to test this change because test cases in the existing test case already make use of the squeeze function for various backends.

Differential Revision: [D33798546](https://our.internmc.facebook.com/intern/diff/D33798546)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
looks great! thanks for the refactor and cleanup!

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33798546).

dzdang committed:
Update on "[Quant][devs] Separated implementations for quantized & non-quantized tensors for the squeeze function"


Summary: This is a PR that address a concern brought up in https://github.com/pytorch/pytorch/issues/54150 regarding opting for separat kernels for quantized & non-quantized tensors, when possible, as opposed to using conditional if-else.

Test plan: Additional testing was not implemented to test this change because test cases in the existing test case already make use of the squeeze function for various backends.

initial

Differential Revision: [D33798546](https://our.internmc.facebook.com/intern/diff/D33798546)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33798546).

dzdang committed:
Update on "[Quant][devs] Separated implementations for quantized & non-quantized tensors for the squeeze function"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR separates the calls to quantized & non-quantized backends for squeeze
using a dispatcher.

Test plan: Additional testing was not implemented to test this change because test cases in the existing test case already make use of the squeeze function for various backends.

initial

Differential Revision: [D33798546](https://our.internmc.facebook.com/intern/diff/D33798546)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33798546).
